lib/mutgen.c

```diff
@@ -25,66 +25,350 @@
+                         1.0/3.0, 1.0/3.0, 1.0/3.0, 0.0};
+static mutation_model_t jukes_cantor_model = {
+    .alphabet_length = 4,
+    .alphabet = acgt_alphabet,
+    .root_distribution = jukes_cantor_model_root_distribution,
+    .mutation_matrix = jukes_cantor_model_mutation_matrix
 };
 
+void
+mutation_model_print_state(mutation_model_t *self, FILE *out)
+{
+    int j, k;
+    double *mutation_row;
+
+    fprintf(out, "mutation_model (%p):: (%d)", (void *) self, self->alphabet_length);
+    fprintf(out, "\talphabet = %.*s\n", self->alphabet_length, self->alphabet);
+    fprintf(out, " distribution =");
+    for (j = 0; j < self->alphabet_length; j++) {
+        fprintf(out, " %0.4f", self->root_distribution[j]);
+    }
+    fprintf(out, "\n\t------------------------------------\t\n");
+    for (j = 0; j < self->alphabet_length; j++) {
+        mutation_row = self->mutation_matrix + j * self->alphabet_length;
+        for (k = 0; k < self->alphabet_length; k++) {
+            fprintf(out, " %0.4f", mutation_row[k]);
+        }
+        fprintf(out, "\n");
+    }
+    fprintf(out, "\n");
+}
+
+static int MSP_WARN_UNUSED
+mutation_model_check_validity(mutation_model_t *self)
+{
+    int ret = 0;
+    int j, k;
+    double prob, min_prob;
+    double *mutation_row;
+
+    /* Check probabilities sum to one */
+    prob = 0;
+    min_prob = 0;
+    for (j = 0; j < self->alphabet_length; j++) {
+        prob += self->root_distribution[j];
+        min_prob = GSL_MIN(min_prob, self->root_distribution[j]);
+    }
+    if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
+        ret = MSP_ERR_BAD_ROOT_PROBABILITIES;
+        goto out;
+    }
+    for (j = 0; j < self->alphabet_length; j++) {
+        prob = 0;
+        min_prob = 0;
+        mutation_row = self->mutation_matrix + j * self->alphabet_length;
+        for (k = 0; k < self->alphabet_length; k++) {
+            prob += mutation_row[k];
+            min_prob = GSL_MIN(min_prob, mutation_row[k]);
+        }
+        if (!doubles_almost_equal(prob, 1.0, 1e-12) || min_prob < 0.0) {
+            ret = MSP_ERR_BAD_MUTATION_MATRIX;
+            goto out;
+        }
+    }
+
+out:
+    return ret;
+}
+
+int MSP_WARN_UNUSED
+mutation_model_alloc(mutation_model_t *self, int alphabet_length,
+        const char *alphabet, double *root_distribution, double *mutation_matrix)
+{
+    int ret = 0;
+    int j, k, jk;
+    char *buff = NULL;
+
+    memset(self, 0, sizeof(mutation_model_t));
+    if (alphabet_length < 2) {
+        ret = MSP_ERR_INSUFFICIENT_ALLELES;
+        goto out;
+    }
+
+    buff = malloc((size_t) alphabet_length * sizeof(*self->alphabet));
+    self->root_distribution = malloc((size_t) alphabet_length * sizeof(*self->root_distribution));
+    self->mutation_matrix = malloc((size_t) (alphabet_length * alphabet_length)
+                                            * sizeof(*self->mutation_matrix));
+    if (self->root_distribution == NULL
```

Comment:
Missing check ``buff == NULL``

Reply:
Why bother with buff? Seems easier to just work with ``self->alphabet`` directly?

Reply:
Probably because I don't understand `const` pointer things: this was the permutation of things that didn't give me errors. =) But: since `self->alphabet` is `const` I can't do this:
```
    for (j = 0; j < alphabet_length; j++) {
        self->alphabet[j] = alphabet[j];
    }
```
and it didn't seem like I should do
```
self->alphabet = alphabet;
```
... beause who knows where alphabet is coming from... but maybe that's ok because `alphabet` is `const`?

Reply:
Right, so I think a lot of the complexity here is because you're statically allocating the basic models. Let's not bother with this, and instead assume the models have been passed in as a parameter.

It'll need a little plumbing, but it's straightforward. I'm happy to do it, if you like. (Possibly quicker, as I'll need to experiment a bit to see what the best way to set things up at the Python C interface)